### PR TITLE
In progress - adding configuration (basic)

### PR DESCRIPTION
**Pull request by gadotroee**

### CODE REVIEW

File: cli/config/config.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	},
+	{
+		CommandLineName:   "uploadInterval",
+		YamlHierarchyName: ConfigurationKeyUploadInterval,
+		DefaultValue:      10,
+		Type:              reflect.Int,
+	},
+	{
+		CommandLineName:   "mizuImage",
+		YamlHierarchyName: ConfigurationKeyMizuImage,
```
Comment (IgorGov) [inline cli/config/config.go:52]: can we remove the prefix "Configuration" from all YamlHierarchyName?

File: cli/config/config.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+func GetConfig() interface{} {
+	return configObj
+}
+
+func MergeAllSettings() error {
+	mizu.Log.Debugf("Merging default values")
+	mergeDefaultValues()
+	mizu.Log.Debugf("Merging settings file values")
+	if err1 := mergeSettingsFileSettings(); err1 != nil {
+		fmt.Printf(mizu.Red, "Invalid settings file\n")
```
Comment (IgorGov) [inline cli/config/config.go:90]: why not using our logger?

Comment (IgorGov) [reply cli/config/config.go:90]: should we print the supplied file path?

Comment (IgorGov) [reply cli/config/config.go:90]: should we print the error itself? to user? just to log file?

File: cli/config/config.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	return "", reflect.Invalid, errors.New(fmt.Sprintf("invalid set argument %s", setFlagKey))
+}
+
+func validateSettingsFileKey(settingsFileKey string) {
+	for _, allowedFlag := range allowedSetFlags {
+		if allowedFlag.YamlHierarchyName == settingsFileKey {
+			return
+		}
+	}
+	fmt.Printf("Invalid settings file. Exit, %v", settingsFileKey)
```
Comment (IgorGov) [inline cli/config/config.go:185]: you already have such print can we merge them?

File: cli/config/config.go (hunk 1 above)
Comment (gadotroee) [reply cli/config/config.go:52]: I think this is a good convention to understand those consts are configuration keys.. maybe I can do it with Enum or something..

File: cli/cmd/tap.go [hunk 4]
```diff
@@ -3,7 +3,9 @@ package cmd
 import (
 	"errors"
 	"fmt"
+	"github.com/romana/rlog"
 	"github.com/spf13/cobra"
+	"github.com/up9inc/mizu/cli/config"
```
[the remark this answers is not in the document]
Comment (gadotroee) [reply cli/cmd/tap.go:7]: yes we can

File: cli/config/config.go (hunk 2 above)
Comment (gadotroee) [reply cli/config/config.go:90]: I think we should print to user because we are going to exit in this case.

Comment (gadotroee) [reply cli/config/config.go:90]: I'm changing to use out logger

File: cli/cmd/config.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	Use:   "config",
+	Short: "Generate example config file to stdout",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		template := mizu.GetTemplateConfig()
+		if outputFileName != "" {
+			data := []byte(template)
+			_ = os.WriteFile(outputFileName, data, 0644)
+			fmt.Println(fmt.Sprintf("Template File written to %s", outputFileName))
+		} else {
+			fmt.Println(template)
```
Comment (IgorGov) [inline cli/cmd/config.go:22]: we should use log here

File: cli/cmd/root.go [hunk 6]
```diff
@@ -1,14 +1,30 @@
...
+	"github.com/up9inc/mizu/cli/mizu"
 )
 
 var rootCmd = &cobra.Command{
 	Use:   "mizu",
 	Short: "A web traffic viewer for kubernetes",
 	Long: `A web traffic viewer for kubernetes
 Further info is available at https://github.com/up9inc/mizu`,
+	PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+		if err := mizu.MergeAllSettings(); err != nil {
```
Comment (IgorGov) [inline cli/cmd/root.go:16]: init config

File: cli/cmd/root.go [hunk 7]
```diff
@@ -1,14 +1,30 @@
...
+			return errors.New(fmt.Sprintf("%v", err))
+		}
+		prettifiedConfig := mizu.GetConfigStr()
+		mizu.Log.Debugf("Final Config: %s", prettifiedConfig)
+		return nil
+	},
+}
+
+func init() {
+	rootCmd.PersistentFlags().StringSliceVar(&mizu.CommandLineValues, "set", []string{}, "Override values using --set")
```
Comment (IgorGov) [inline cli/cmd/root.go:27]: should we have "CommandLineValues" in current file

File: cli/mizu/config.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+	val, err := strconv.Atoi(stringVal)
+	if err != nil {
+		Log.Warningf("Invalid value %v for key %s", val, key)
+		os.Exit(1)
+	}
+	return val
+}
+
+func InitSettings(commandLineValues []string) error {
```
Comment (IgorGov) [inline cli/mizu/config.go:71]: InitConfig

File: cli/mizu/config.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+func mergeDefaultValues() {
+	for _, allowedFlag := range allowedSetFlags {
+		Log.Debugf("Setting %v to %v", allowedFlag.YamlHierarchyName, allowedFlag.DefaultValue)
+		configObj[allowedFlag.YamlHierarchyName] = allowedFlag.DefaultValue
+	}
+}
+
+func mergeSettingsFileSettings() error {
```
Comment (IgorGov) [inline cli/mizu/config.go:117]: mergeConfigFile

File: cli/mizu/config.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+func flagFromAllowed(setFlagKey string) (string, reflect.Kind, error) {
+	for _, allowedFlag := range allowedSetFlags {
+		if strings.ToLower(allowedFlag.CommandLineName) == strings.ToLower(setFlagKey) {
+			return allowedFlag.YamlHierarchyName, allowedFlag.Type, nil
+		}
+	}
+	return "", reflect.Invalid, errors.New(fmt.Sprintf("invalid set argument %s", setFlagKey))
+}
+
+func validateSettingsFileKey(settingsFileKey string) {
```
Comment (IgorGov) [inline cli/mizu/config.go:186]: validateConfigFileKey

**Review by IgorGov — DISMISSED** [review]

**Review by IgorGov — APPROVED** [review]

**Pull request merged**